packages/contracts/src/contracts/current/protocol/Exchange/MixinWrapperFunctions.sol

```diff
@@ -16,281 +16,259 @@
-            mstore(add(start, 68), add(orderAddresses, 64))    // makerToken
-            mstore(add(start, 100), add(orderAddresses, 96))   // takerToken
-            mstore(add(start, 132), add(orderAddresses, 128))  // feeRecipient
-            mstore(add(start, 164), orderValues)            // makerTokenAmount
-            mstore(add(start, 196), add(orderValues, 32))   // takerTokenAmount
-            mstore(add(start, 228), add(orderValues, 64))   // makerFee
-            mstore(add(start, 260), add(orderValues, 96))   // takerFee
-            mstore(add(start, 292), add(orderValues, 128))  // expiration
+            // Load free memory pointer
+            let start := mload(0x40)
+
+            // Write function signature
+            mstore(start, fillOrderSelector)
+
+            // Write order struct
+            mstore(add(start, 4), mload(order))             // makerAddress
+            mstore(add(start, 36), mload(add(order, 32)))   // takerAddress
+            mstore(add(start, 68), mload(add(order, 64)))   // makerTokenAddress
+            mstore(add(start, 100), mload(add(order, 96)))  // takerTokenAddress
+            mstore(add(start, 132), mload(add(order, 128))) // feeRecipientAddress
+            mstore(add(start, 164), mload(add(order, 160))) // makerTokenAmount
+            mstore(add(start, 196), mload(add(order, 192))) // takerTokenAmount
+            mstore(add(start, 228), mload(add(order, 224))) // makerFeeAmount
+            mstore(add(start, 260), mload(add(order, 256))) // takerFeeAmount
+            mstore(add(start, 292), mload(add(order, 288))) // expirationTimeSeconds
+            mstore(add(start, 324), mload(add(order, 320))) // salt
+
+            // Write takerTokenFillAmount
             mstore(add(start, 356), takerTokenFillAmount)
+
+            // Write signature offset
             mstore(add(start, 388), 416)
-            mstore(add(start, 420), signatureLength)
-        }
-        
-        // Write signature contents and padding one byte at a time
-        for (uint256 i = 0; i < signatureLength; i++) {
-            input[452 + i] = signature[i];
-        }
-        for (i = 0; i < paddingLength; i++) {
-            input[452 + signatureLength + i] = 0x00;
-        }
-        
-        // Call the function
-        assembly {
-            success := delegatecall(
-                gas,       // If fillOrder `revert()`s, we recover unused gas
-                address,   // call this contract
-                start,     // pointer to start of input
-                inputSize, // length of input
-                start,     // store output over input
-                32         // output is 32 bytes
+
+            // Write signature length
+            let sigLen := mload(signature)
+            mstore(add(start, 420), sigLen)
+
+            // Calculate signature length with padding
+            let paddingLen := sub(32, mod(sigLen, 32))
```

Comment:
When `sigLen` is an exact multiple of 32, it will add 32 bytes of padding. In that case it should add no padding.

`let paddingLen := mod(sub(0, sigLen), 32)` should do the trick.